src/window/mod.rs

```diff
@@ -0,0 +1,601 @@
+//!
+//! Architecture
+//! ------------
+//!
+//! Windowing is based around two core traits and one generic core
+//! operator: [`Clock`], [`Windower`], and
+//! [`StatefulWindowUnary::stateful_window_unary`].
+//!
+//! All user-facing windowing operators like
+//! [`crate::dataflow::Dataflow::reduce_window`] are built on top of
+//! [`StatefulWindowUnary::stateful_window_unary`] so we only have to
+//! write the window management code once.
+//!
+//! That operator itself is based upon
+//! [`crate::operators::StatefulUnary`], which provides a generic
+//! abstraction to the recovery system. We get recovery for "free" as
+//! long as we play by the rules of that operator.
+//!
+//! This system follows our standard pattern of having parallel Python
+//! config objects and Rust impl structs for each trait of behavior we
+//! want. E.g. [`SystemClockConfig`] represents a token in Python for
+//! how to create a [`SystemClock`].
+
+use crate::dataflow::StepId;
+use crate::operators::StatefulUnary;
+use crate::pyo3_extensions::{StateKey, TdPyAny, TdPyCallable};
+use crate::recovery::StateBackup;
+use crate::with_traceback;
+use chrono::{Duration, NaiveDateTime};
+use log::debug;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::task::Poll;
+use timely::dataflow::{Scope, Stream};
+use timely::{Data, ExchangeData};
+
+/// Base class for a clock config.
+///
+/// This describes how a windowing operator should determine the
+/// current time and the time for each element.
+///
+/// Use a specific subclass of this that matches the time definition
+/// you'd like to use.
+#[pyclass(module = "bytewax.window", subclass)]
+#[pyo3(text_signature = "()")]
+pub(crate) struct ClockConfig;
+
+/// Use the system time inside the windowing operator to determine
+/// times.
+///
+/// If the dataflow has no more input, all windows are closed.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `clock_config` parameter to
+///     your windowing operator.
+#[pyclass(module="bytewax.window", extends=ClockConfig)]
+struct SystemClockConfig {}
+
+#[pymethods]
+impl SystemClockConfig {
+    #[new]
+    #[args()]
+    fn new() -> (Self, ClockConfig) {
+        (Self {}, ClockConfig {})
+    }
+
+    /// Pickle as a tuple.
+    fn __getstate__(&self) -> (&str,) {
+        ("SystemClockConfig",)
+    }
+
+    /// Unpickle from tuple of arguments.
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("SystemClockConfig",)) = state.extract() {
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for SystemClockConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+pub(crate) fn build_clock<V>(
+    py: Python,
+    clock_config: Py<ClockConfig>,
+) -> Result<Box<dyn Clock<V>>, String> {
+    let window_config = clock_config.as_ref(py);
```

Comment:
Should this var be `clock_config`? I'm not seeing where it's used for windowing here (yet)